### PR TITLE
[FLINK-22936] Support column comment in Schema and ResolvedSchema

### PR DESCRIPTION
**Pull request by hackergin**
## What is the purpose of the change

*Support column comment in Schema and ResolvedSchema*

## Brief change log

  - Support column comment in Schema and ResolvedSchema

## Verifying this change

This change is already covered by existing tests
 such as
 -  testSchemaResolution  in SchemaResolutionTest .
 -  testUnresolvedSchemaString  in SchemaResolutionTest .
 -  testResolvedSchemaString  in SchemaResolutionTest .
 
I only made small change in SchemaResolutionTest. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java [hunk 1]
```diff
@@ -221,6 +221,15 @@ public Builder fromColumns(List<UnresolvedColumn> unresolvedColumns) {
             return this;
         }
 
+        /** Apply comment to the previous column. */
+        public Builder withComment(@Nullable String comment) {
+            if (columns.size() > 0) {
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java:226]: Please throw an `IllegalArgumentException` if there is no preceding column defined.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java [hunk 2]
```diff
@@ -665,13 +687,31 @@ public int hashCode() {
             this.dataType = dataType;
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java:687]: We can share single constructor implementation by calling `this(columnName, dataType, null)`.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Column.java [hunk 3]
```diff
@@ -45,9 +45,18 @@
...
 
+    protected final @Nullable String comment;
+
     private Column(String name, DataType dataType) {
         this.name = name;
         this.dataType = dataType;
+        this.comment = null;
+    }
+
+    private Column(String name, DataType dataType, @Nullable String comment) {
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Column.java:56]: Ditto. Please share constructor implementation.

**Review by wuchong — COMMENTED** [review]
Thanks for the contribution. I left some comments.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java [hunk 4]
```diff
@@ -620,15 +631,27 @@ private void addResolvedConstraint(UniqueConstraint constraint) {
     /** Super class for all kinds of columns in an unresolved schema. */
     public abstract static class UnresolvedColumn {
         final String columnName;
+        final @Nullable String comment;
 
         UnresolvedColumn(String columnName) {
```
Comment (hackergin) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java:636]: @wuchong  I found this constructor is never used now , should I  remove this ? .

Comment (wuchong) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java:636]: Sure.

**Comment by hackergin** [comment]
@flinkbot  run azure

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java [hunk 5]
```diff
@@ -226,6 +226,8 @@ public Builder withComment(@Nullable String comment) {
             if (columns.size() > 0) {
                 columns.set(
                         columns.size() - 1, columns.get(columns.size() - 1).withComment(comment));
+            } else {
+                throw new IllegalArgumentException("There is no preceding column defined.");
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java:230]:
```suggestion
                throw new IllegalArgumentException("Method \"withComment\" must be followed by a column definition, but there is no preceding column defined.");
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Column.java [hunk 6]
```diff
@@ -48,9 +48,7 @@
     protected final @Nullable String comment;
 
     private Column(String name, DataType dataType) {
-        this.name = name;
-        this.dataType = dataType;
-        this.comment = null;
+        this(name, dataType, null);
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Column.java:51]: This constructor can be removed now.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java [hunk 7]
```diff
@@ -632,8 +634,7 @@ private void addResolvedConstraint(UniqueConstraint constraint) {
         final @Nullable String comment;
 
         UnresolvedColumn(String columnName) {
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java:636]: This method can be removed now.

**Comment by hackergin** [comment]
@wuchong  thanks for your review , I updated the code.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Column.java [hunk 8]
```diff
@@ -108,6 +119,13 @@ public String asSummaryString() {
                             sb.append(" ");
                             sb.append(e);
                         });
+        getComment()
+                .ifPresent(
+                        c -> {
+                            sb.append(" COMMENT '");
+                            sb.append(c);
```
Comment (Airblader) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Column.java:126]: We should wrap this with `EncodingUtils#escapeSingleQuotes`.

Comment (hackergin) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Column.java]: @Airblader  thanks for the review, I update the code, and add the related test case.

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**